okhttp/src/main/java/okhttp3/internal/cache/CacheInterceptor.kt

```diff
@@ -0,0 +1,267 @@
+   */
+  @Throws(IOException::class)
+  private fun cacheWritingResponse(cacheRequest: CacheRequest?, response: Response): Response {
+    // Some apps return a null body; for compatibility we treat that like a null cache request.
+    if (cacheRequest == null) return response
+    val cacheBodyUnbuffered = cacheRequest.body()
+
+    val source = response.body()!!.source()
+    val cacheBody = cacheBodyUnbuffered.buffer()
+
+    val cacheWritingSource = object : Source {
+      var cacheRequestClosed: Boolean = false
+
+      @Throws(IOException::class)
+      override fun read(sink: Buffer, byteCount: Long): Long {
+        val bytesRead: Long
+        try {
+          bytesRead = source.read(sink, byteCount)
+        } catch (e: IOException) {
+          if (!cacheRequestClosed) {
+            cacheRequestClosed = true
+            cacheRequest.abort() // Failed to write a complete cache response.
+          }
+          throw e
+        }
+
+        if (bytesRead == -1L) {
+          if (!cacheRequestClosed) {
+            cacheRequestClosed = true
+            cacheBody.close() // The cache response is complete!
+          }
+          return -1
+        }
+
+        sink.copyTo(cacheBody.buffer, sink.size - bytesRead, bytesRead)
+        cacheBody.emitCompleteSegments()
+        return bytesRead
+      }
+
+      override fun timeout(): Timeout {
+        return source.timeout()
+      }
+
+      @Throws(IOException::class)
+      override fun close() {
+        if (!cacheRequestClosed && !discard(this, ExchangeCodec.DISCARD_STREAM_TIMEOUT_MILLIS,
+                MILLISECONDS)) {
+          cacheRequestClosed = true
+          cacheRequest.abort()
+        }
+        source.close()
+      }
+    }
+
+    val contentType = response.header("Content-Type")
+    val contentLength = response.body()?.contentLength() ?: -1
+    return response.newBuilder()
+        .body(RealResponseBody(contentType, contentLength, cacheWritingSource.buffer()))
+        .build()
+  }
+
+  companion object {
+
+    private fun stripBody(response: Response?): Response? {
+      if (response?.body() != null)
+        return response.newBuilder().body(null).build()
+      else
+        return response
+    }
+
+    /** Combines cached headers with a network headers as defined by RFC 7234, 4.3.4.  */
+    private fun combine(cachedHeaders: Headers, networkHeaders: Headers): Headers {
+      val result = Headers.Builder()
+
+      for (index in cachedHeaders.names().indices) {
+        val fieldName = cachedHeaders.name(index)
+        val value = cachedHeaders.value(index)
+        if ("Warning".equals(fieldName, ignoreCase = true) && value.startsWith("1")) {
+          continue
+        }
+        if (isContentSpecificHeader(fieldName) ||
+            !isEndToEnd(fieldName) ||
+            networkHeaders[fieldName] == null) {
+          addHeaderLenient(result, fieldName, value)
+        }
+      }
+
+      for (index in networkHeaders.names().indices) {
```

Comment:
does the `.indices` call allocate?

Reply:
> does the `.indices` call allocate?

I tried to decompile `for (index in cachedHeaders.names().indices)` into a class file and got:
```
Int i = 0;
For (int j = ((Collection)cachedHeaders.names()).size(); i < j; index++)

```
It does not seem to allocate extra memory to the `IntRange` object.

There is a blog about this issue, it seems that Kotlin is doing very well:
[Exploring Kotlin’s hidden costs](https://medium.com/@BladeCoder/exploring-kotlins-hidden-costs-part-3-3bf6e0dbf0a4)
In this article, you can directly search for the keyword **collection indices** and you can see related information.